app/api/companies.py

```diff
@@ -138,43 +142,89 @@ def get_company(id):
 def add_company():
     data = request.get_json() or {}
 
-    # Validation
-    if 'company_id' in data:
-        return bad_request("Create company cannot include company_id. For modifying existing companies please use the PUT method")
-    if 'company_name' not in data:
-        return bad_request("Must include company_name field")
-    if Companies.query.filter_by(company_name=data['company_name']).first():
-        return bad_request("A company with that name already exists, please use another name")
-
-    company = Companies()
-    # TODO: Finalize the from_dict method
-    company.from_dict_new(data)
-    db.session.add(company)
-    db.session.commit()
-
-    response = jsonify(company.to_dict())
+    try:
+        result = CompaniesValidationSchema().load(data)
+    except ValidationError as err:
+        print(err.messages)
+        print(err.valid_data)
+        return bad_request(err.messages)
+
+    # TODO: Validate if company name is already in DB?
+
+    new_company = Companies(company_name=result['company_name'], logo_image_src=result['logo_image_src'],
+                            website=result['website'], year=result['year'], company_size=result['company_size'])
+
+    # Check if city is already in DB:
+    city = Cities.query.filter_by(
+        city_name=result['city_name'].title()).first()
+    if city is not None:
+        new_company.city_id = city.city_id
+        db.session.add(new_company)
+        db.session.commit()
+    else:
+        new_city = Cities(
+            city_name=result['city_name'].title(), region=result['region'])
+        new_city.company.append(new_company)
+        db.session.add(new_city)
+        db.session.commit()
+
+    # Insert Meta Data:
+    insert_meta(result['disciplines'], 'disciplines', new_company.company_id)
+    insert_meta(result['branches'], 'branches', new_company.company_id)
+    insert_meta(result['tags'], 'tags', new_company.company_id)
+
+    response = jsonify(new_company.to_dict())
     response.status_code = 201
     response.headers['Location'] = url_for(
-        'api.get_company', id=company.company_id)
+        'api.get_company', id=new_company.company_id)
 
     cache.clear()
     return response
 
 
 # TODO: Create a specific error message when selecting a non-existing company_id
-@bp.route('/v1/companies/<int:id>', methods=['PUT'])
+@bp.route('/v1/companies/<int:id>', methods=['PATCH'])
 @jwt_required()
 def update_company(id):
-    company = Companies.query.get_or_404(id)
+
     data = request.get_json() or {}
-    # TODO: Validation
-    if False:
-        return bad_request('false')
-    # TODO: Finalize the from_dict method
-    company.from_dict(data, new_company=False)
-    db.session.commit()
+
+    try:
+        result = CompaniesPatchSchema().load(data, partial=True)
```

Comment:
Nice - with `partial=True` we should even be able to update the record when sending a subselection of all fields (as should be possible with PATCH)

Reply:
Yes, that works as expected.